File: app/src/main/java/org/ametro/ng/ui/fragments/MapListFragment.java
package org.ametro.ng.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.AsyncTaskLoader;
import androidx.loader.content.Loader;
import androidx.appcompat.widget.SearchView;

import org.ametro.ng.R;
import org.ametro.ng.app.ApplicationEx;
import org.ametro.ng.catalog.entities.MapCatalog;
import org.ametro.ng.catalog.entities.MapInfo;
import org.ametro.ng.ui.adapters.MapListAdapter;
import org.ametro.ng.ui.loaders.ExtendedMapInfo;
import org.ametro.ng.ui.loaders.ExtendedMapStatus;

public class MapListFragment extends Fragment implements
        SearchView.OnQueryTextListener,
        SearchView.OnCloseListener,
        LoaderManager.LoaderCallbacks<MapCatalog>,
        AdapterView.OnItemClickListener,
        AbsListView.MultiChoiceModeListener,
        View.OnClickListener {

    private MapListAdapter adapter;
    private ProgressBar progressBar;
    private TextView progressText;
    private View noMapsView;
    private View emptyView;

    private String filterValue;

    private MapCatalog localMapCatalog;
    private ListView list;
    private static final int LOCAL_CATALOG_LOADER = 1;

    private IMapListEventListener listener = new IMapListEventListener() {
        @Override
        public void onOpenMap(MapInfo map) { }

        @Override
        public void onLoadedMaps(ExtendedMapInfo[] maps) { }
    };

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_map_list_view, container, false);

        progressBar = view.findViewById(R.id.progressBar);
        progressText = view.findViewById(R.id.progressText);

        noMapsView = view.findViewById(R.id.no_maps);
        noMapsView.setOnClickListener(this);

        emptyView = view.findViewById(R.id.empty);

        list = view.findViewById(R.id.list);
        list.setOnItemClickListener(this);

        adapter = new MapListAdapter(getActivity(), ApplicationEx.getInstance(getActivity()).getCountryFlagProvider());
        list.setAdapter(adapter);
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        LoaderManager.getInstance(this).initLoader(LOCAL_CATALOG_LOADER, null, this).forceLoad();
    }

    public void setMapListEventListener(IMapListEventListener newListener) {
        listener = newListener;
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        listener.onOpenMap(adapter.getItem(position));
    }

    @NonNull
    @Override
    public Loader<MapCatalog> onCreateLoader(int id, Bundle args) {
        ApplicationEx app = ApplicationEx.getInstance(requireActivity());
        return new MapCatalogAsyncTaskLoader(requireContext(), app);
    }

    @Override
    public void onLoadFinished(@NonNull Loader<MapCatalog> loader, MapCatalog data) {
        if (data == null) {
            return;
        }
        localMapCatalog = data;
        resetAdapter();
    }

    @Override
    public void onLoaderReset(@NonNull Loader<MapCatalog> loader) { }

    @Override
    public boolean onClose() {
        filterValue = null;
        adapter.getFilter().filter(null);
        return true;
    }

    @Override
    public boolean onQueryTextSubmit(String s) {
        filterValue = s;
        adapter.getFilter().filter(s);
        return true;
    }

    @Override
    public boolean onQueryTextChange(String s) {
        filterValue = s;
        adapter.getFilter().filter(s);
        return true;
    }

    private void setNoMapsShown() {
        list.setEmptyView(null);
        progressText.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
        list.setVisibility(View.GONE);
        emptyView.setVisibility(View.GONE);
        noMapsView.setVisibility(View.VISIBLE);
    }

    private void setListShown() {
        progressText.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
        noMapsView.setVisibility(View.GONE);
        emptyView.setVisibility(View.VISIBLE);
        list.setVisibility(View.VISIBLE);
        list.setEmptyView(emptyView);
    }

    @Override
    public void onItemCheckedStateChanged(ActionMode mode, int position, long id, boolean checked) {
        final int checkedCount = list.getCheckedItemCount();
        mode.setTitle(checkedCount + " " + getText(R.string.msg_selected));
        adapter.toggleSelection(position);
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        return false;
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        MenuInflater inflater = mode.getMenuInflater();
        inflater.inflate(R.menu.map_list_context_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        return true;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        adapter.clearSelection();
    }

    @Override
    public void onClick(View v) {
        // do nothing
    }

    public interface IMapListEventListener {
        void onOpenMap(MapInfo map);
        void onLoadedMaps(ExtendedMapInfo[] maps);
    }

    private void resetAdapter() {
        if (localMapCatalog == null || localMapCatalog.getMaps().length == 0) {
            adapter.clear();
            adapter.getFilter().filter(filterValue);
            setNoMapsShown();
            return;
        }

        MapInfo[] localMaps = localMapCatalog.getMaps();
        ExtendedMapInfo[] maps = new ExtendedMapInfo[localMaps.length];
        for (int i = 0; i < maps.length; i++) {
            maps[i] = new ExtendedMapInfo(localMaps[i],ExtendedMapStatus.Installed);
        }

        adapter.clear();
        adapter.addAll(maps);
        adapter.getFilter().filter(filterValue);

        setListShown();
        listener.onLoadedMaps(maps);
    }

    private static class MapCatalogAsyncTaskLoader extends AsyncTaskLoader<MapCatalog> {
        private final ApplicationEx app;

        public MapCatalogAsyncTaskLoader(Context context, ApplicationEx app) {
            super(context);
            this.app = app;
        }

        @Override
        public MapCatalog loadInBackground() {
            return app.getMapCatalogProvider().getMapCatalog();
        }
    }

}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/ui/fragments/MapListFragment.kt
package io.github.romangolovanov.apps.ametro.ui.fragments

import android.content.Context
import android.os.Bundle
import android.view.ActionMode
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.ListView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.loader.app.LoaderManager
import androidx.loader.content.AsyncTaskLoader
import androidx.loader.content.Loader
import io.github.romangolovanov.apps.ametro.R
import io.github.romangolovanov.apps.ametro.app.ApplicationEx
import io.github.romangolovanov.apps.ametro.catalog.entities.MapCatalog
import io.github.romangolovanov.apps.ametro.catalog.entities.MapInfo
import io.github.romangolovanov.apps.ametro.ui.adapters.MapListAdapter
import io.github.romangolovanov.apps.ametro.ui.loaders.ExtendedMapInfo
import io.github.romangolovanov.apps.ametro.ui.loaders.ExtendedMapStatus

class MapListFragment : Fragment(),
    SearchView.OnQueryTextListener,
    SearchView.OnCloseListener,
    LoaderManager.LoaderCallbacks<MapCatalog>,
    AdapterView.OnItemClickListener,
    AbsListView.MultiChoiceModeListener,
    View.OnClickListener {

    private lateinit var adapter: MapListAdapter
    private lateinit var progressBar: ProgressBar
    private lateinit var progressText: TextView
    private lateinit var noMapsView: View
    private lateinit var emptyView: View
    private lateinit var list: ListView

    private var filterValue: String? = null
    private var localMapCatalog: MapCatalog? = null

    private var listener: IMapListEventListener = object : IMapListEventListener {
        override fun onOpenMap(map: MapInfo) {}
        override fun onLoadedMaps(maps: Array<ExtendedMapInfo>) {}
    }

    companion object {
        private const val LOCAL_CATALOG_LOADER = 1
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        @Nullable container: ViewGroup?,
        @Nullable savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_map_list_view, container, false)

        progressBar = view.findViewById(R.id.progressBar)
        progressText = view.findViewById(R.id.progressText)

        noMapsView = view.findViewById(R.id.no_maps)
        noMapsView.setOnClickListener(this)

        emptyView = view.findViewById(R.id.empty)

        list = view.findViewById(R.id.list)
        list.setOnItemClickListener(this)

        adapter = MapListAdapter(requireActivity(), ApplicationEx.getInstance(requireActivity()).countryFlagProvider)
        list.adapter = adapter
        return view
    }

    override fun onResume() {
        super.onResume()
        LoaderManager.getInstance(this).initLoader(LOCAL_CATALOG_LOADER, null, this).forceLoad()
    }

    fun setMapListEventListener(newListener: IMapListEventListener) {
        listener = newListener
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        listener.onOpenMap(adapter.getItem(position)!!)
    }

    @NonNull
    override fun onCreateLoader(id: Int, args: Bundle?): Loader<MapCatalog> {
        val app = ApplicationEx.getInstance(requireActivity())
        return MapCatalogAsyncTaskLoader(requireContext(), app)
    }

    override fun onLoadFinished(@NonNull loader: Loader<MapCatalog>, data: MapCatalog?) {
        if (data == null) return
        localMapCatalog = data
        resetAdapter()
    }

    override fun onLoaderReset(@NonNull loader: Loader<MapCatalog>) {}

    override fun onClose(): Boolean {
        filterValue = null
        adapter.filter.filter(null)
        return true
    }

    override fun onQueryTextSubmit(s: String): Boolean {
        filterValue = s
        adapter.filter.filter(s)
        return true
    }

    override fun onQueryTextChange(s: String): Boolean {
        filterValue = s
        adapter.filter.filter(s)
        return true
    }

    private fun setNoMapsShown() {
        list.emptyView = null
        progressText.visibility = View.GONE
        progressBar.visibility = View.GONE
        list.visibility = View.GONE
        emptyView.visibility = View.GONE
        noMapsView.visibility = View.VISIBLE
    }

    private fun setListShown() {
        progressText.visibility = View.GONE
        progressBar.visibility = View.GONE
        noMapsView.visibility = View.GONE
        emptyView.visibility = View.VISIBLE
        list.visibility = View.VISIBLE
        list.emptyView = emptyView
    }

    override fun onItemCheckedStateChanged(mode: ActionMode, position: Int, id: Long, checked: Boolean) {
        val checkedCount = list.checkedItemCount
        mode.title = "$checkedCount ${getText(R.string.msg_selected)}"
        adapter.toggleSelection(position)
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean = false

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        val menuInflater: MenuInflater = mode.menuInflater
        menuInflater.inflate(R.menu.map_list_context_menu, menu)
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean = true

    override fun onDestroyActionMode(mode: ActionMode) {
        adapter.clearSelection()
    }

    override fun onClick(v: View) {
        // do nothing
    }

    interface IMapListEventListener {
        fun onOpenMap(map: MapInfo)
        fun onLoadedMaps(maps: Array<ExtendedMapInfo>)
    }

    private fun resetAdapter() {
        val catalog = localMapCatalog
        if (catalog == null || catalog.maps.isEmpty()) {
            adapter.clear()
            adapter.filter.filter(filterValue)
            setNoMapsShown()
            return
        }

        val localMaps = catalog.maps
        val maps = Array(localMaps.size) { i -> ExtendedMapInfo(localMaps[i], ExtendedMapStatus.Installed) }

        adapter.clear()
        adapter.addAll(*maps)
        adapter.filter.filter(filterValue)

        setListShown()
        listener.onLoadedMaps(maps)
    }

    private class MapCatalogAsyncTaskLoader(context: Context, private val app: ApplicationEx) :
        AsyncTaskLoader<MapCatalog>(context) {

        override fun loadInBackground(): MapCatalog {
            return app.mapCatalogProvider.mapCatalog
        }
    }
}
